contributors/devel/sig-architecture/vendor.md

```diff
@@ -0,0 +1,172 @@
+updated. Valid reasons include:
+ 1. We need new functionality that is in a later version.
+ 2. New or improved APIs in the dependency significantly improve Kubernetes code.
+ 3. Bugs were fixed that impact Kubernetes.
+ 4. Security issues were fixed even if they don't impact Kubernetes yet.
+ 5. Performance, scale, or efficiency was meaningfully improved.
+ 6. We need dependency A and there is a transitive dependency B.
+ 7. Kubernetes has an older level of a dependency that is precluding being able
+to work with other projects in the ecosystem.
+
+## Theory of operation
+
+The `go.mod` file in the root of `k8s.io/kubernetes` describes dependencies using two directives:
+
+* `require` directives list the preferred version of dependencies (this is auto-updated by go tooling to the maximum preferred version of the module)
+* `replace` directives pin to specific tags or commits
+
+## Adding or updating a dependency
+
+The most common things people need to do with deps are add and update them.
+These operations are handled the same way:
+
+For the sake of examples, consider that we have discovered a wonderful Go
+library at `example.com/go/frob`.
+
+Step 1: Ensure there is go code in place that references the packages you want to use.
+```go
+import "example.com/go/frob"
+...
+frob.DoStuff()
+```
+
+Step 2: Determine what version of the dependency you want to use, and add that version to the go.mod file:
+
+```sh
+hack/pin-dependency.sh example.com/go/frob v1.0.4
+```
+
+This fetches the dependency, resolves the specified sha or tag, and adds two entries to the `k8s.io/kubernetes` `go.mod` file:
+
+```
+require (
+    example.com/go/frob v1.0.4
+    ...
+)
+
+replace (
+    example.com/go/frob => example.com/go/frob v1.0.4
+    ...
+)
+```
+
+The `require` directive indicates our module requires `example.com/go/frob` >= `v1.0.4`.
+If our module was included as a dependency in a build with other modules that also required `example.com/go/frob`,
+the maximum required version would be selected (unless the main module in that build pinned to a lower version).
+
+The `replace` directive pins us to the desired version when running go commands within kubernetes/kubernetes.
+
+Step 3: Rebuild the `vendor` directory and update the `go.mod` files for all staging repositories:
+```sh
+hack/update-vendor.sh
+```
+
+Step 4: Check if the new dependency requires newer versions of existing dependencies we have pinned.
+You can check this by:
+1. running `hack/lint-dependencies.sh` against your branch and against `master` and comparing the results
+2. checking if any new `replace` directives were added to `go.mod` files of components inside the staging directory.
+
+Staging components with `replace` directives are the most problematic, because consumers of those components
+will use different versions of libraries than the ones we build and test Kubernetes with by default.
+
+If transitive dependencies need to be updated as a result of the new dependency,
```

Comment:
I'm a little confused about why the tools don't topologically sort and do this in a single pass?

Reply:
open to improving the tools, just trying to get the primitive operations in place first